### PR TITLE
NIAD-2789: Handle inline attachments from EMIS where the description contains _only_ the filename.

### PR DESCRIPTION
**Pull request by adrianclay**
Example description from EMIS:

```xml
   <eb:Description xml:lang="en">EB8E92C9-7346-4A0A-ACFA-A0BAC0713441_1-MB-DOC.doc</eb:Description>
```

### CODE REVIEW

**Comment by adrianclay** [comment]
Posting the latest stack trace here for easier debugging.

```logs
2023-08-07 14:19:41 2023-08-07 13:19:41.759 Level=ERROR Logger=u.n.a.p.t.t.MhsQueueMessageHandler ConversationId=6D8B4944-C0EF-4381-8ACD-F88BFFE6B244 Thread="org.springframework.jms.JmsListenerEndpointContainer#0-1" Message="Unable to parse Ebxml References"
2023-08-07 14:19:41 java.text.ParseException: Unable to parse originalFilename
2023-08-07 14:19:41     at uk.nhs.adaptors.pss.translator.util.XmlParseUtilService.parseFilename(XmlParseUtilService.java:101)
2023-08-07 14:19:41     at uk.nhs.adaptors.pss.translator.task.EhrExtractMessageHandler.buildPatientAttachmentLogFromAttachment(EhrExtractMessageHandler.java:247)
2023-08-07 14:19:41     at uk.nhs.adaptors.pss.translator.task.EhrExtractMessageHandler.processInternalAttachmentsAndReturnSkeletonLog(EhrExtractMessageHandler.java:153)
2023-08-07 14:19:41     at uk.nhs.adaptors.pss.translator.task.EhrExtractMessageHandler.handleMessage(EhrExtractMessageHandler.java:106)
2023-08-07 14:19:41     at uk.nhs.adaptors.pss.translator.task.MhsQueueMessageHandler.handleMessage(MhsQueueMessageHandler.java:79)
2023-08-07 14:19:41     at uk.nhs.adaptors.pss.translator.amqp.MhsQueueConsumer.receive(MhsQueueConsumer.java:38)
2023-08-07 14:19:41     at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2023-08-07 14:19:41     at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
2023-08-07 14:19:41     at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2023-08-07 14:19:41     at java.base/java.lang.reflect.Method.invoke(Method.java:568)
2023-08-07 14:19:41     at org.springframework.messaging.handler.invocation.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:169)
2023-08-07 14:19:41     at org.springframework.messaging.handler.invocation.InvocableHandlerMethod.invoke(InvocableHandlerMethod.java:119)
2023-08-07 14:19:41     at org.springframework.jms.listener.adapter.MessagingMessageListenerAdapter.invokeHandler(MessagingMessageListenerAdapter.java:114)
2023-08-07 14:19:41     at org.springframework.jms.listener.adapter.MessagingMessageListenerAdapter.onMessage(MessagingMessageListenerAdapter.java:77)
2023-08-07 14:19:41     at org.springframework.jms.listener.AbstractMessageListenerContainer.doInvokeListener(AbstractMessageListenerContainer.java:736)
2023-08-07 14:19:41     at org.springframework.jms.listener.AbstractMessageListenerContainer.invokeListener(AbstractMessageListenerContainer.java:696)
2023-08-07 14:19:41     at org.springframework.jms.listener.AbstractMessageListenerContainer.doExecuteListener(AbstractMessageListenerContainer.java:674)
2023-08-07 14:19:41     at org.springframework.jms.listener.AbstractPollingMessageListenerContainer.doReceiveAndExecute(AbstractPollingMessageListenerContainer.java:331)
2023-08-07 14:19:41     at org.springframework.jms.listener.AbstractPollingMessageListenerContainer.receiveAndExecute(AbstractPollingMessageListenerContainer.java:270)
2023-08-07 14:19:41     at org.springframework.jms.listener.DefaultMessageListenerContainer$AsyncMessageListenerInvoker.invokeListener(DefaultMessageListenerContainer.java:1237)
2023-08-07 14:19:41     at org.springframework.jms.listener.DefaultMessageListenerContainer$AsyncMessageListenerInvoker.executeOngoingLoop(DefaultMessageListenerContainer.java:1227)
2023-08-07 14:19:41     at org.springframework.jms.listener.DefaultMessageListenerContainer$AsyncMessageListenerInvoker.run(DefaultMessageListenerContainer.java:1120)
2023-08-07 14:19:41     at java.base/java.lang.Thread.run(Thread.java:833)

```

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/task/EhrExtractMessageHandler.java [hunk 1]
```diff
@@ -262,6 +287,29 @@ private PatientAttachmentLog buildPatientAttachmentLogFromAttachment(
     private PatientAttachmentLog buildPatientAttachmentLogFromExternalAttachment(
```
Comment (adrianclay) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/task/EhrExtractMessageHandler.java:287]: We _shouldn't_ need this behaviour for an `ExternalAttachment` (based on the example we've been given within the ticket).

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/task/EhrExtractMessageHandler.java [hunk 2]
```diff
@@ -242,6 +244,29 @@ private PatientAttachmentLog buildPatientAttachmentLogFromAttachment(
         PatientMigrationRequest migrationRequest,
         InboundMessage.Attachment attachment) throws ParseException {
 
+        String description = attachment.getDescription();
```
Comment (adrianclay) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/task/EhrExtractMessageHandler.java:247]: It's a shame we have this behaviour duplicated across both this class and `InlineAttachment`.

Comment (adrianclay) [reply gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/task/EhrExtractMessageHandler.java:247]: I'm a little surprised that we don't have any unit tests which needed changing for `EhrExtractMessageHandler` class.

File: gp2gp-translator/src/integrationTest/java/uk/nhs/adaptors/pss/translator/LargeMessagingIT.java [hunk 3]
```diff
@@ -53,6 +53,16 @@ public void handleUk06WithCidAttachment() throws JSONException {
         verifyBundle("/json/LargeMessage/expectedBundleScenario1.json");
     }
 
+    @Test
+    public void handleUk06WithEMISCidAttachment() throws JSONException {
```
Comment (adrianclay) [inline gp2gp-translator/src/integrationTest/java/uk/nhs/adaptors/pss/translator/LargeMessagingIT.java:57]: `LargeMessaging` has a specific meaning within GP2GP, so I'm a little unhappy that this test lives within `LargeMessagingIT.java`.

Though I think this issue already exists, so perhaps the file needs renaming.

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/task/EhrExtractMessageHandler.java [hunk 4]
```diff
@@ -242,6 +244,29 @@ private PatientAttachmentLog buildPatientAttachmentLogFromAttachment(
...
+        Matcher matcher = pattern.matcher(description);
+        Boolean isEmis = matcher.find();
+
+        if(isEmis){
+            return PatientAttachmentLog.builder()
+                    .mid(messageId)
+                    .filename(attachment.getDescription())
+                    .parentMid(null)
+                    .patientMigrationReqId(migrationRequest.getId())
+                    .contentType("plain/text")
```
Comment (adrianclay) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/task/EhrExtractMessageHandler.java:258]: This is hard-coded which is suspicious.

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/task/EhrExtractMessageHandler.java [hunk 5]
```diff
@@ -242,6 +244,29 @@ private PatientAttachmentLog buildPatientAttachmentLogFromAttachment(
...
+                    .compressed(false)
+                    .largeAttachment(false)
```
Comment (adrianclay) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/task/EhrExtractMessageHandler.java:260]: Should we be worried about this?

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/task/EhrExtractMessageHandler.java (hunk 2 above)
Comment (adrianclay) [reply gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/task/EhrExtractMessageHandler.java:255]: After doing some digging, I suspect that moving all the logic into `InlineAttachment` feels like a sensible route forward for reducing the duplication.

File: gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/service/AttachmentHandlerServiceStoreAttachmentTests.java [hunk 6]
```diff
@@ -148,6 +135,25 @@ public void When_ValidListOfAttachmentsAndConversationIdIsGiven_Expect_DoesNotTh
         verify(storageManagerService, times(2)).uploadFile(any(), any(), any());
     }
 
+    @Test
+    public void When_ValidListOfAttachmentsFromEMIS_Expect_DoesNotThrow() throws ValidationException,
```
[the remark this answers is not in the document]
Comment (adrianclay) [reply gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/service/AttachmentHandlerServiceStoreAttachmentTests.java:139]: We tried writing a test to see what happened when a malformed Description was used for a COPC message, and the app didn't seem to care either with or without our changes.

File: gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/task/EhrExtractMessageHandlerTest.java [hunk 7]
```diff
@@ -480,6 +490,54 @@ public void When_HandleSingleMessageWithSkeletonAttachmentAndValidDataIsCalled_E
...
+        assertEquals("test.txt", log.getFilename());
+        assertNull(log.getParentMid());
+        assertEquals(MIGRATION_REQUEST_ID, log.getPatientMigrationReqId());
+        assertEquals(attachment.getContentType(), log.getContentType());
+        assertFalse(log.getCompressed());
+        assertFalse(log.getLargeAttachment());
+        assertTrue(log.getOriginalBase64()); // _Should_ this be true?
+        assertFalse(log.getSkeleton());
+        assertTrue(log.getUploaded());
+        assertEquals(0, log.getLengthNum()); // Zero length seems sus
```
Comment (adrianclay) [inline gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/task/EhrExtractMessageHandlerTest.java:536]: Will this cause issues?

File: gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/task/EhrExtractMessageHandlerTest.java [hunk 8]
```diff
@@ -480,6 +490,54 @@ public void When_HandleSingleMessageWithSkeletonAttachmentAndValidDataIsCalled_E
...
+        verify(patientAttachmentLogService, times(1)).addAttachmentLog(patientAttachmentLogCaptor.capture());
+        PatientAttachmentLog log = patientAttachmentLogCaptor.getValue();
+        assertEquals("MESSAGE-ID", log.getMid());
+        assertEquals("test.txt", log.getFilename());
+        assertNull(log.getParentMid());
+        assertEquals(MIGRATION_REQUEST_ID, log.getPatientMigrationReqId());
+        assertEquals(attachment.getContentType(), log.getContentType());
+        assertFalse(log.getCompressed());
+        assertFalse(log.getLargeAttachment());
+        assertTrue(log.getOriginalBase64()); // _Should_ this be true?
```
Comment (adrianclay) [inline gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/task/EhrExtractMessageHandlerTest.java:533]: Is this correct?

Comment (adrianclay) [reply gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/task/EhrExtractMessageHandlerTest.java:533]: I _don't think_ this field is used for anything, but probably safer to make it false?

File: gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/task/EhrExtractMessageHandlerTest.java (hunk 7 above)
Comment (Alex-Nita) [reply gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/task/EhrExtractMessageHandlerTest.java:536]: Yes, you're correct. It needs to be `attachment.getPayload().length()`

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/task/EhrExtractMessageHandler.java (hunk 2 above)
Comment (adrianclay) [reply gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/task/EhrExtractMessageHandler.java:255]: I've had another look at this, and I'm not sure how we could refactor it into a nice way. I suspect the reason for this is I don't fully understand the requriements.  @benhession might be able to help 😢

File: gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/task/EhrExtractMessageHandlerTest.java (hunk 7 above)
Comment (benhession) [reply gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/task/EhrExtractMessageHandlerTest.java:536]: I'm not a fan of the length checks. The COPC message handler will check the length if it has been given in the description. Setting it to 0 will bypass it, so I think it's right, it should be 0.

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/task/EhrExtractMessageHandler.java (hunk 2 above)
Comment (benhession) [reply gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/task/EhrExtractMessageHandler.java:255]: I've left some comments below that relate to this, we could use the `XMLParseUtilService` to reduce the regex duplication. The attachment log is keeping track of the attachments that are expected. The `InlineAttachment` class is used to represent an attachment or part of an attachment that has been received.

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/util/XmlParseUtilService.java [hunk 9]
```diff
@@ -37,6 +37,10 @@ public class XmlParseUtilService {
 
     private final XPathService xPathService;
 
+    private static String filenamePattern = "Filename=\"([\\S]+[A-Za-z\\d\\-_. ]*)\"";
```
[the remark this answers is not in the document]
Comment (Alex-Nita) [reply gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/util/XmlParseUtilService.java:40]: This expression would match the rest of the string as well, including the other fields from the description. The committed one doesn't.

**Review by benhession — APPROVED** [review]

**Pull request merged**